pkg/scheduler/scheduling_queue.go

```diff
@@ -499,6 +499,37 @@ func (sq *SchedulingQueue) getHeadRoom() *resources.Resource {
 	return resources.ComponentWiseMin(headRoom, parentHeadRoom)
 }
 
+// this function returns the max headRoom of a queue
+// this doesn't get the partition resources into the consideration
+func (sq *SchedulingQueue) getMaxHeadRoom() *resources.Resource {
```

Comment:
The getMaxHeadRoom and getHeadRoom share most of the code.
Can we simplify this by taking the recursive code out for the two different calculations:
```
func (sq *SchedulingQueue) getHeadRoom() *resources.Resource {
	var parentHeadRoom *resources.Resource
	if sq.parent != nil {
		parentHeadRoom = sq.parent.getHeadRoom()
	}
	return sq.internalHeadRoom(parentHeadRoom)
}

func (sq *SchedulingQueue) getMaxHeadRoom() *resources.Resource {
	var parentHeadRoom *resources.Resource
	if sq.parent != nil {
		parentHeadRoom = sq.parent.getMaxHeadRoom()
	} else {
		return nil
	}
	return sq.internalHeadRoom(parentHeadRoom)
}
```

And then in the `internalHeadRoom` the real calculation:
```
func (sq *SchedulingQueue) internalHeadRoom(parentHeadRoom *resources.Resource) *resources.Resource {
	sq.RLock()
	defer sq.RUnlock()
	headRoom := sq.QueueInfo.GetMaxResource()

	// if we have no max set headroom is always the same as the parent
	if headRoom == nil {
		return parentHeadRoom
	}
	// calculate unused
	headRoom.SubFrom(sq.allocating)
	headRoom.SubFrom(sq.QueueInfo.GetAllocatedResource())
	// check the minimum of the two: parentHeadRoom is nil for root
	if parentHeadRoom == nil {
		return headRoom
	}
	return resources.ComponentWiseMin(headRoom, parentHeadRoom)
}
```

Reply:
Yeah. this is brilliant, I just updated the code like this.